File: amr/sim/simulator.py
"""Top-level 2D simulator: advances the robot, emits noisy odometry and scans.

Each step integrates the ground-truth robot, derives a noisy odometry increment
(in the robot frame) and accumulates it into a drifting odom pose, and emits a
lidar scan every cfg.lidar.scan_every steps.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

import numpy as np

from amr.core.config import AmrConfig
from amr.core.geometry import pose_compose, wrap_angle
from amr.core.types import LaserScan, Pose2D, Twist2D
from amr.sim.lidar import Lidar
from amr.sim.robot import DiffDriveRobot
from amr.sim.world import World


@dataclass
class SimStepResult:
    ground_truth: Pose2D
    odom_pose: Pose2D
    odom_delta: Pose2D
    scan: Optional[LaserScan]
    collided: bool
    sim_time: float


def _arc_robot_frame(v: float, w: float, dt: float) -> Pose2D:
    """Exact-arc integration of (v, w) over dt, expressed in the robot frame
    (start pose at the origin, heading +x). Returns Pose2D(dx_fwd, dy_left, dth).
    """
    if abs(w) < 1e-9:
        return Pose2D(v * dt, 0.0, 0.0)
    R = v / w
    dth = w * dt
    dx = R * math.sin(dth)
    dy = R * (1.0 - math.cos(dth))
    return Pose2D(dx, dy, wrap_angle(dth))


class Simulator:
    def __init__(self, world: World, cfg: AmrConfig, rng: np.random.Generator):
        self.world = world
        self.cfg = cfg
        self.rng = rng
        self.time = 0.0
        self._steps = 0
        self.robot = DiffDriveRobot(cfg.robot, world.spawn)
        self.lidar = Lidar(cfg.lidar, rng)
        self.odom_pose = Pose2D(world.spawn.x, world.spawn.y, world.spawn.theta)

    def step(self, cmd: Twist2D) -> SimStepResult:
        dt = self.cfg.sim.dt

        # Advance ground truth.
        self.robot.step(cmd, dt, self.world)
        self._steps += 1
        self.time += dt

        # Noisy odometry from the actual (clamped) velocity.
        v = self.robot.vel.v
        w = self.robot.vel.omega
        nz = self.cfg.sim.odom_noise
        v_std = nz.alpha_v * abs(v) + nz.floor
        w_std = nz.alpha_w * abs(w) + nz.floor
        v_meas = v + self.rng.normal(0.0, v_std)
        w_meas = w + self.rng.normal(0.0, w_std)

        odom_delta = _arc_robot_frame(v_meas, w_meas, dt)
        self.odom_pose = pose_compose(self.odom_pose, odom_delta)

        # Scan every scan_every steps.
        scan = None
        scan_every = self.cfg.lidar.scan_every
        if scan_every > 0 and (self._steps % scan_every == 0):
            scan = self.lidar.scan(self.world, self.robot.pose, self.time)

        return SimStepResult(
            ground_truth=Pose2D(self.robot.pose.x, self.robot.pose.y,
                                self.robot.pose.theta),
            odom_pose=Pose2D(self.odom_pose.x, self.odom_pose.y,
                             self.odom_pose.theta),
            odom_delta=odom_delta,
            scan=scan,
            collided=self.robot.collided,
            sim_time=self.time,
        )

File: src/amr_sim/src/simulator.cpp
// SPDX-License-Identifier: Apache-2.0
#include "amr_sim/simulator.hpp"

#include <cmath>
#include <utility>

#include "amr_core/geometry.hpp"

namespace amr_sim {

amr_core::Pose2D arc_robot_frame(double v, double w, double dt) {
  if (std::abs(w) < 1e-9) {
    return amr_core::Pose2D{v * dt, 0.0, 0.0};
  }
  const double R = v / w;
  const double dth = w * dt;
  const double dx = R * std::sin(dth);
  const double dy = R * (1.0 - std::cos(dth));
  return amr_core::Pose2D{dx, dy, amr_core::wrap_angle(dth)};
}

Simulator::Simulator(World world, const amr_core::AmrConfig& cfg,
                     std::mt19937& rng)
    : world_(std::move(world)),
      cfg_(cfg),
      rng_(rng),
      time_(0.0),
      steps_(0),
      robot_(cfg_.robot, world_.spawn()),
      lidar_(cfg_.lidar, rng_),
      odom_pose_(world_.spawn()) {}

SimStepResult Simulator::step(const amr_core::Twist2D& cmd) {
  const double dt = cfg_.sim.dt;

  // Advance ground truth.
  robot_.step(cmd, dt, world_);
  ++steps_;
  time_ += dt;

  // Noisy odometry from the actual (clamped) velocity.
  const double v = robot_.vel().v;
  const double w = robot_.vel().omega;
  const amr_core::OdomNoiseConfig& nz = cfg_.sim.odom_noise;
  const double v_std = nz.alpha_v * std::abs(v) + nz.floor;
  const double w_std = nz.alpha_w * std::abs(w) + nz.floor;
  std::normal_distribution<double> v_dist(0.0, v_std);
  std::normal_distribution<double> w_dist(0.0, w_std);
  const double v_meas = v + v_dist(rng_);
  const double w_meas = w + w_dist(rng_);

  const amr_core::Pose2D odom_delta = arc_robot_frame(v_meas, w_meas, dt);
  odom_pose_ = amr_core::pose_compose(odom_pose_, odom_delta);

  SimStepResult result;
  result.ground_truth = robot_.pose();
  result.odom_pose = odom_pose_;
  result.odom_delta = odom_delta;
  result.collided = robot_.collided();
  result.sim_time = time_;

  // Scan every scan_every steps.
  const int scan_every = cfg_.lidar.scan_every;
  if (scan_every > 0 && (steps_ % scan_every == 0)) {
    result.scan = lidar_.scan(world_, robot_.pose(), time_);
    result.scan_ready = true;
  } else {
    result.scan_ready = false;
  }

  return result;
}

}  // namespace amr_sim
